odoo/models.py

```diff
@@ -1250,8 +1254,11 @@ def _extract_records(self, fields_, data, log=lambda a: None, limit=float('inf')
             if fields[fnames[0]].type != 'one2many'
         ])
         # Checks if the provided row has any non-empty one2many fields
-        def only_o2m_values(row):
-            return any(get_o2m_values(row)) and not any(get_nono2m_values(row))
+        # or if the non-relational fields match and the relational fields don't
+        def row_continues(row, index_row_o2m_vals=None, index_row_nono2m_vals=None):
+            o2m_vals = get_o2m_values(row)
+            nono2m_vals = get_nono2m_values(row)
+            return (any(o2m_vals) and not any(nono2m_vals)) or (nono2m_vals == index_row_nono2m_vals and o2m_vals != index_row_o2m_vals)
```

Comment:
Where are the unit tests that validate those extra use-cases?

Reply:
Added a few test cases, left some comments for a few more
@xmo-odoo @odony @rco-odoo I know this task is postponed but it'd be really cool if you guys could guide me to completion when you have time